rfcs/2020-11-02-remap-strict-return-value-types.md

```diff
@@ -0,0 +1,296 @@
+This example program, when executed, always resolves to a `String` value.
+
+Another example program is `floor(.bar)`. This program again consists of two
+expressions:
+
+1. The query path expression `.bar`, which resolves to any value the `bar`
+   object field contains.
+2. The function expression `floor`, which resolves to either an `Integer`, or
+   `Float`.
+
+Since the `floor` function is the final expression that runs in the program, the
+outcome of a _successful execution_ of the program is either an `Integer` or
+`Float`.
+
+### expression trait
+
+Expressions are implemented using the `Expression` trait:
+
+```rust
+pub trait Expression: Send + Sync {
+    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>>;
+}
+```
+
+Expressions can be fallible (not covered in this RFC), can return `None` (not
+covered in-depth in this RFC), or return a resolved `Value`.
+
+### defining the expected resolved values
+
+To allow the execution context of a program to determine the outcome of running
+the program, we extend the `Expression` trait as follows:
+
+```rust
+pub trait Expression: Send + Sync {
+    fn execute(&self, state: &mut State, object: &mut dyn Object) -> Result<Option<Value>>;
+
+    /// Describes which [`ValueKind`]s the expression can resolve to.
+    ///
+    /// bike-shedding of function and enum names welcome.
+    fn resolves_to(&self) -> ResolveKind;
+}
+```
+
+We define `ResolveKind` as follows:
+
+```rust
+pub enum ResolveKind {
+    /// The expression can resolve to any value at runtime.
+    Any,
+
+    /// The expressions resolves to one of the defined [`ValueKind`]s.
+    OneOf(Vec<ValueKind>),
+
+    /// If the expression succeeds, it might resolve to a value, but doesn't
+    /// have to.
+    Maybe(Box<ResolveKind>),
+}
+```
+
+The `ResolveKind` enum allows us to determine if an expression resolves to a
+predefined list of values. It also defines if the expression can resolve to
+"nothing" (e.g. `None`).
+
+`ValueKind` is defined as follows:
+
+```rust
+pub enum ValueKind {
+    String,
+    Integer,
+    Float,
+    Boolean,
+    Timestamp,
+    Map,
+    Array,
+}
+```
+
+As can be seen from the `ValueKind`, it is not possible to define the value
+kinds contained within a map or array, only the top-level value kind of an
+expression can be defined.
+
+### transient expressions
+
+Some expressions don't have their own return value, but instead resolve to the
+return value of an expression they themselves execute.
+
+One such an example is the `IfStatement` expression, which resolves to either
+the `true_expression` or the `false_expression`.
+
+Because expressions are initialized at compile-time, we are able to have such
+"transient expressions" return the values they resolve to based on whatever
+expressions they themselves hold:
+
+```rust
+impl Expression for IfStatement {
+    fn resolves_to(&self) -> ResolveKind {
+        let true_resolves = self.false_expression.resolves_to();
```

Comment:
Should be `self.true_expression.resolves_to()`?

Reply:
Nice one, you caught my read-the-RFC Easter egg. If I were a betting man, I would have put my money on you in the first place, but I'm not, so instead of sharing my profits, I'll give you some internet fame to enjoy 😄 👍